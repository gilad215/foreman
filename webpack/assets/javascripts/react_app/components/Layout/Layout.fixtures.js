const mockOnClick = jest.fn();

// PF-React Component Data

const PFitems = [{
  title: 'Monitor', initialActive: true, iconClass: 'fa fa-tachometer', subItems: subItemsA, href: '/',
},
{
  title: 'Hosts', initialActive: false, iconClass: 'fa fa-server', subItems: subItemsB, href: '/',
},
{
  title: 'Configure', initialActive: false, iconClass: 'fa fa-wrench', subItems: subItemsC, href: '/',
}];

const subItemsA = [{
  title: 'A', isDivider: false, onClick: mockOnClick,
},
{
  title: 'B', isDivider: true, onClick: mockOnClick,
},
{
  title: 'C', isDivider: false, onClick: mockOnClick,
},
];
const subItemsB = [{
  title: 'D', isDivider: false, onClick: mockOnClick,
},
{
  title: 'E', isDivider: true, onClick: mockOnClick,
},
{
  title: 'F', isDivider: false, onClick: mockOnClick,
},
];
const subItemsC = [{
  title: 'G', isDivider: false, onClick: mockOnClick,
},
{
  title: 'H', isDivider: true, onClick: mockOnClick,
},
{
  title: 'I', isDivider: false, onClick: mockOnClick,
},
];

// Server Hash Data
const monitorChildren = [{
  type: 'item', name: 'Dashboard', url: '/',
},
{ type: 'divider', name: 'Hello' },
{
  type: 'item', name: 'Facts', url: '/fact_values',
},
];

const hostsChildren = [{
  type: 'item', name: 'All Hosts', url: '/hosts/new',
},
{ type: 'divider', name: 'There' },
{
  type: 'item', name: 'Architectures', url: '/architectures',
},
];

const userChildren = [{
  type: 'item', name: 'Environments', url: '/environments',
},
{ type: 'divider', name: 'Dear' },
{
  type: 'item', name: 'Architectures', url: '/architectures',
},
];
const infrastructureChildren = [{
  type: 'item', name: 'Domains', url: '/domains',
},
{ type: 'divider', name: 'Friend' },
{
  type: 'item', name: 'Realms', url: '/realms',
},
];

const hashItemsA = [{
  type: 'sub_menu', name: 'Monitor', icon: 'fa fa-tachometer', children: monitorChildren, active: false,
},
{
  type: 'sub_menu', name: 'Hosts', icon: 'fa fa-server', children: hostsChildren, active: false,
}];

const hashItemsB = [{
  type: 'sub_menu', name: 'User', icon: 'fa fa-wrench', children: userChildren, active: false,
},
{
  type: 'sub_menu', name: 'Infrastructure', icon: 'pficon pficon-network', children: infrastructureChildren, active: false,
}];

export const serverItems = [...hashItemsA, ...hashItemsB];
// DATA PROPS

const logo = '/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg';
const user = {
  current_user: {
    user: {
      id: 4, login: 'admin', firstname: 'Admin', lastname: 'User',
    },
  },
  user_dropdown: [
    {
      type: 'sub_menu', name: 'User', icon: 'fa fa-user', children: subItemsA,
    },
  ],
};

const organizations = {
  current_org: {
    organization: { id: 1, name: 'org1' },
  },
  available_organizations: [{ id: 1, name: 'org1', href: '/organizations/1-org1/select' }],
};

const locations = {
  current_location: {
    location: { id: 1, name: 'yaml' },
  },
  available_locations: [{ id: 1, name: 'yaml', href: '/locations/1-yaml/select' }],
};

const serverUser = {
  current_user: {
    user: {
      firstname: 'G',
      lastname: 'L',
    },
  },
  user_dropdown: [{
    children: [{
      type: 'item', url: '/', name: 'My Account',
    }, { type: 'divider' },
    ],
  }],
};

export const layoutMock = {
  items: PFitems,
  data: {
    menu: [hashItemsA, hashItemsB],
    locations,
    organizations,
    root: '/',
    logo,
    notification_url: '/notification_recipients',
    taxonomies: { locations: true, organizations: true },
    user,
  },
};

export const noItemsMock = {
  ...layoutMock,
  items: [],
};

export const hasTaxonomiesMock = {
  ...layoutMock,
  currentLocation: 'yaml',
  currentOrganization: 'org1',
};

export const taxonomySwitcherProps = {
  organizations: organizations.available_organizations,
  locations: locations.available_locations,
  taxonomiesBool: { locations: true, organizations: true },
  isLoading: true,
};

export const userDropdownProps = {
  user: serverUser,
  notification_url: '/',
};

