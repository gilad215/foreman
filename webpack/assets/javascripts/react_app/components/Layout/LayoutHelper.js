import {
  changeOrganization,
  changeLocation,
  navigateTo,
} from '../../../foreman_navigation';

export const getCurrentPath = () => window.location.pathname;

export const combineMenuItems = (data) => {
  const items = [];

  data.menu.forEach((menu) => {
    menu.forEach((item) => {
      items.push(item);
    });
  });
  if (data.taxonomies.organizations) {
    items.push(createOrgItem(data.organizations.available_organizations));
  }

  if (data.taxonomies.locations) {
    items.push(createLocationItem(data.locations.available_locations));
  }
  return items;
};

const createOrgItem = (orgs) => {
  const anyOrg = {
    name: 'Any Organization',
    onClick: () => {
      changeOrganization('Any Organization');
      navigateTo('/organizations/clear');
    },
  };
  const childrenArray = [];
  childrenArray.push(anyOrg);

  orgs.forEach((org) => {
    const childObject = {
      type: org.type,
      name: org.title,
      onClick: () => {
        navigateTo(org.href);
        changeOrganization(org.title);
      },
      url: org.href,
    };
    childrenArray.push(childObject);
  });

  const orgItem = {
    type: 'sub_menu',
    name: 'Organizations',
    icon: 'fa fa-building',
    children: childrenArray,
    className: 'visible-xs-block',
    active: false,
  };
  return orgItem;
};

const createLocationItem = (locations) => {
  const anyLoc = {
    name: 'Any Location',
    onClick: () => {
      changeLocation('Any Location');
      navigateTo('/locations/clear');
    },
  };
  const childrenArray = [];
  childrenArray.push(anyLoc);

  locations.forEach((loc) => {
    const childObject = {
      type: loc.type,
      name: loc.title,
      onClick: () => {
        navigateTo(loc.href);
        changeLocation(loc.title);
      },
      url: loc.href,
    };
    childrenArray.push(childObject);
  });

  const locItem = {
    type: 'sub_menu',
    name: 'Locations',
    icon: 'fa fa-globe',
    children: childrenArray,
    className: 'visible-xs-block',
    active: false,
  };
  return locItem;
};
