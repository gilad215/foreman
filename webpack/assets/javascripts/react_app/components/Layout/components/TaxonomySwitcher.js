import React from 'react';
import PropTypes from 'prop-types';
import { Nav, Spinner } from 'patternfly-react';
import NavItem from './NavItem';
import { navigateTo } from '../../../../foreman_navigation';

const TaxonomySwitcher = ({
  currentOrganization,
  currentLocation,
  organizations,
  locations,
  taxonomiesBool,
  onOrgClick,
  onLocationClick,
  isLoading,
  ...props
}) => (
  <Nav navbar pullLeft className="navbar-iconic" {...props}>
    {taxonomiesBool.organizations && (
      <NavItem className="dropdown org-switcher" id="organization-dropdown">
        <a
          href="#"
          className="dropdown-toggle nav-item-iconic"
          data-toggle="dropdown"
        >
          {currentOrganization}
          <span className="caret" />
        </a>
        <ul className="dropdown-menu">
          <li className="dropdown-header">{__('Organization')}</li>
          <li>
            <a
              className="organizations_clear"
              onClick={() => {
                onOrgClick(__('Any Organization'));
                navigateTo('/organizations/clear');
              }}
            >
              {__('Any Organizations')}
            </a>
          </li>
          <li>
            <a
              className="manage-menu"
              className="organizations"
              href="/organizations"
            >
              {__('Manage Organizations')}
            </a>
          </li>
          <li className="divider" />
          {organizations.map((organization, i) => (
            <li key={i}>
              <a
                className="org_menuitem"
                data-id={`aid_organizations_${organization.title}`}
                onClick={() => {
                  onOrgClick(organization.title);
                  navigateTo(organization.href);
                }}
              >
                {organization.title}
              </a>
            </li>
          ))}
        </ul>
      </NavItem>
    )}
    {taxonomiesBool.locations && (
      <NavItem className="dropdown org-switcher" id="location-dropdown">
        <a
          href="#"
          className="dropdown-toggle nav-item-iconic"
          data-toggle="dropdown"
        >
          {currentLocation}
          <span className="caret" />
        </a>
        <ul className="dropdown-menu">
          <li className="dropdown-header">{__('Location')}</li>
          <li>
            <a
              className="locations_clear"
              onClick={() => {
                onLocationClick(__('Any Location'));
                navigateTo('/locations/clear');
              }}
            >
              {__('Any Location')}
            </a>
          </li>
          <li>
            <a
              className="manage-menu"
              className="locations"
              href="/locations"
            >
              {__('Manage Locations')}
            </a>
          </li>
          <li className="divider" />
          {locations.map((location, i) => (
            <li key={i}>
              <a
                className="location_menuitem"
                data-id={`aid_locations_${location.title}`}
                onClick={() => {
                  onLocationClick(location.title);
                  navigateTo(location.href);
                }}
              >
                {location.title}
              </a>
            </li>
          ))}
        </ul>
      </NavItem>
    )}
    {isLoading && (
      <NavItem id="vertical-spinner">
        <Spinner size="md" inverse loading />
      </NavItem>
    )}
  </Nav>
);
TaxonomySwitcher.propTypes = {
  /** Additional element css classes */
  className: PropTypes.string,
  /** Locations array */
  locations: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    href: PropTypes.string,
  })),
  /** Organizations array */
  organizations: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    href: PropTypes.string,
  })),
  /** isLoading Prop */
  isLoading: PropTypes.bool,
  /** current Organization */
  currentOrganization: PropTypes.string,
  /** current Location */
  currentLocation: PropTypes.string,
};
TaxonomySwitcher.defaultProps = {
  className: '',
  locations: [],
  organizations: [],
  isLoading: false,
  currentLocation: '',
  currentOrganization: '',
};
export default TaxonomySwitcher;
