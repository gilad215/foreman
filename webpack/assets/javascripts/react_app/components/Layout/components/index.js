import React from 'react';
import { Loadable } from '../../../utils';

const loadingLayout = () => (
  <div id="layout">
    <div className="navbar navbar-pf-vertical-loading" />
    <div className="nav-pf-vertical nav-pf-vertical-with-sub-menus nav-pf-persistent-secondary" />
  </div>
);

export default Loadable(
  {
    loader: () => import('./Layout'),
  },
  loadingLayout
);
