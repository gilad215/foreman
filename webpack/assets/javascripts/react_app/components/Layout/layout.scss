@import '../../common/colors.scss';

.nav-pf-secondary-nav {
  background-color: $nav-pf-vertical-active-bg-color !important;

  .mobile-active {
    // Mark active Location/Org in Mobile view
    a {
      .list-group-item-value {
        background-color: $nav-pf-vertical-secondary-active-bg-color !important;
      }
    }
  }

  .list-group-item {
    a {
      background-color: $nav-pf-vertical-active-bg-color !important;

      &:hover {
        .list-group-item-value {
          background-color: $nav-pf-vertical-secondary-active-bg-color !important;
        }
      }
    }

    span {
      color: $nav-pf-vertical-active-icon-color !important;
    }
  }
}

.nav-pf-vertical {
  background-color: $nav-pf-vertical-secondary-active-bg-color !important;

  .list-group {
    .list-group-item {
      a {
        span {
          color: $nav-pf-vertical-icon-color;
        }

        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }

        &::after {
          color: $nav-pf-vertical-icon-color;
        }
      }

      &:hover {
        a {
          background-color: $nav-pf-vertical-active-bg-color !important;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }

      &.active {
        a {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }
    }
  }
}

#account_menu {
  outline: none;
}

#loading-div {
  height: 0;
}

.loading-bar {
  background-color: #39a5dc;
  height: 3px;
  position: absolute;
}

.navbar-header {
  width: 221.22px;
}
