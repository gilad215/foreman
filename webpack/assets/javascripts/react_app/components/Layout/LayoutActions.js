
import { isEmpty } from 'lodash';
import { combineMenuItems } from './LayoutHelper';
import { selectMenuItems } from './LayoutSelectors';
import {
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_UPDATE_ITEMS,
  LAYOUT_CHANGE_LOCATION,
  LAYOUT_CHANGE_ORG,
} from './LayoutConstants';

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeActiveMenu = primary => (dispatch, getState) => {
  const menuItems = selectMenuItems(getState()).map(item =>
    (primary.title === item.name ? { ...item, active: true } : { ...item, active: false }));

  dispatch({
    type: LAYOUT_UPDATE_ITEMS,
    payload: {
      menuItems: isEmpty(menuItems) ? [] : menuItems,
      active: primary.title,
    },
  });
};

export const fetchMenuItems = data => (dispatch) => {
  const menuItems = combineMenuItems(data);
  dispatch({
    type: LAYOUT_UPDATE_ITEMS,
    payload: {
      menuItems: isEmpty(menuItems) ? [] : menuItems,
    },
  });
};

export const changeOrganization = org => (dispatch) => {
  dispatch({
    type: LAYOUT_CHANGE_ORG,
    payload: {
      org,
    },
  });
};

export const changeLocation = location => (dispatch) => {
  dispatch({
    type: LAYOUT_CHANGE_LOCATION,
    payload: {
      location,
    },
  });
};
