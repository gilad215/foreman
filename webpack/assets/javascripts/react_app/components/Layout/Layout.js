import React from 'react';
import PropTypes from 'prop-types';

import { VerticalNav } from 'patternfly-react';
import { isEmpty } from 'lodash';
import { noop } from '../../common/helpers';

import TaxonomySwitcher from './components/TaxonomySwitcher';
import UserDropdowns from './components/UserDropdowns';
import './layout.scss';

class Layout extends React.Component {
  componentDidMount() {
    const {
      items,
      data,
      fetchMenuItems,
      changeLocation,
      currentLocation,
      changeOrganization,
      currentOrganization,
    } = this.props;
    if (items.length === 0) fetchMenuItems(data);

    if (!isEmpty(data.locations.current_location)
     && currentLocation !== data.locations.current_location.location.name) {
      changeLocation(data.locations.current_location.location.name);
    }

    if (!isEmpty(data.organizations.current_org)
     && currentOrganization !== data.organizations.current_org.organization.name) {
      changeOrganization(data.organizations.current_org.organization.name);
    }
  }

  render() {
    const {
      items,
      data,
      isLoading,
      changeActiveMenu,
      changeOrganization,
      changeLocation,
      currentOrganization,
      currentLocation,
    } = this.props;
    return (
      <VerticalNav
        hoverDelay={0}
        items={items}
        onItemClick={changeActiveMenu}
        {...this.props}
      >
        <VerticalNav.Masthead>
          <VerticalNav.Brand
            title="FOREMAN"
            iconImg={data.logo}
            href={data.root}
          />
          <TaxonomySwitcher
            taxonomiesBool={data.taxonomies}
            currentLocation={currentLocation}
            locations={data.locations.available_locations}
            onLocationClick={changeLocation}
            currentOrganization={currentOrganization}
            organizations={data.organizations.available_organizations}
            onOrgClick={changeOrganization}
            isLoading={isLoading}
          />
          <UserDropdowns
            notificationUrl={data.notification_url}
            user={data.user}
            changeActiveMenu={changeActiveMenu}
          />
        </VerticalNav.Masthead>
      </VerticalNav>
    );
  }
}

Layout.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    className: PropTypes.string,
    iconClass: PropTypes.string,
    initialActive: PropTypes.bool,
    subItems: PropTypes.arrayOf(PropTypes.shape({
      title: PropTypes.string,
      isDivider: PropTypes.bool,
      className: PropTypes.string,
      onClick: PropTypes.func,
    })),
  })),
  data: PropTypes.shape({
    menu: PropTypes.array,
    locations: PropTypes.object,
    organizations: PropTypes.object,
    root: PropTypes.string,
    logo: PropTypes.string,
    notification_url: PropTypes.string,
    taxonomies: PropTypes.object,
    user: PropTypes.object,
  }),
  currentOrganization: PropTypes.string,
  currentLocation: PropTypes.string,
  isLoading: PropTypes.bool,
  fetchMenuItems: PropTypes.func,
  changeActiveMenu: PropTypes.func,
  changeOrganization: PropTypes.func,
  changeLocation: PropTypes.func,
};

Layout.defaultProps = {
  items: [],
  data: {},
  currentOrganization: 'Any Organization',
  currentLocation: 'Any Location',
  isLoading: false,
  fetchMenuItems: noop,
  changeActiveMenu: noop,
  changeOrganization: noop,
  changeLocation: noop,
};

export default Layout;
