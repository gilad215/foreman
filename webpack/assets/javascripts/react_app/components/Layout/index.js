import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './LayoutActions';
import reducer from './LayoutReducer';
import { patternflyMenuItemsSelector } from './LayoutSelectors';

import Layout from './Layout';

// map state to props
const mapStateToProps = state => ({
  items: patternflyMenuItemsSelector(state),
  isLoading: state.layout.isLoading,
  activeMenu: state.layout.activeMenu,
  currentOrganization: state.layout.currentOrganization,
  currentLocation: state.layout.currentLocation,
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(Layout);
