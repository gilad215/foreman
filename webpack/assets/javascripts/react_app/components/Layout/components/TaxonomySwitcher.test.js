import React from 'react';
import { shallow } from 'enzyme';
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

import TaxonomySwitcher from '../components/TaxonomySwitcher';
import { taxonomySwitcherProps } from '../Layout.fixtures';

const createStubs = () => ({
  onOrgClick: jest.fn(),
  onLocationClick: jest.fn(),
});

const fixtures = {
  'render switcher w/loading': { ...taxonomySwitcherProps, ...createStubs() },
};

describe('TaxonomySwitcher', () => {
  describe('rendering', () => testComponentSnapshotsWithFixtures(TaxonomySwitcher, fixtures));

  describe('trigger onClicks', () => {
    const wrapper = shallow(<TaxonomySwitcher { ...taxonomySwitcherProps } { ...createStubs() } />);

    wrapper.find('.organizations_clear').simulate('click');
    wrapper.find('.org_menuitem').simulate('click');
    wrapper.find('.locations_clear').simulate('click');
    wrapper.find('.location_menuitem').simulate('click');
  });
});
