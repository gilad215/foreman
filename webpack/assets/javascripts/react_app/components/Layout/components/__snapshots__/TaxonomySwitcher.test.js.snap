// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxonomySwitcher rendering render switcher w/loading 1`] = `
<Nav
  bsClass="nav"
  className=""
  justified={false}
  navbar={true}
  pullLeft={true}
  pullRight={false}
  stacked={false}
>
  <NavItem
    activeHref=""
    activeKey=""
    className="dropdown org-switcher"
    id="organization-dropdown"
  >
    <a
      className="dropdown-toggle nav-item-iconic"
      data-toggle="dropdown"
      href="#"
    >
      <span
        className="caret"
      />
    </a>
    <ul
      className="dropdown-menu"
    >
      <li
        className="dropdown-header"
      >
        Organization
      </li>
      <li>
        <a
          className="organizations_clear"
          onClick={[Function]}
        >
          Any Organizations
        </a>
      </li>
      <li>
        <a
          className="organizations"
          href="/organizations"
        >
          Manage Organizations
        </a>
      </li>
      <li
        className="divider"
      />
      <li
        key="0"
      >
        <a
          className="org_menuitem"
          data-id="aid_organizations_undefined"
          onClick={[Function]}
        />
      </li>
    </ul>
  </NavItem>
  <NavItem
    activeHref=""
    activeKey=""
    className="dropdown org-switcher"
    id="location-dropdown"
  >
    <a
      className="dropdown-toggle nav-item-iconic"
      data-toggle="dropdown"
      href="#"
    >
      <span
        className="caret"
      />
    </a>
    <ul
      className="dropdown-menu"
    >
      <li
        className="dropdown-header"
      >
        Location
      </li>
      <li>
        <a
          className="locations_clear"
          onClick={[Function]}
        >
          Any Location
        </a>
      </li>
      <li>
        <a
          className="locations"
          href="/locations"
        >
          Manage Locations
        </a>
      </li>
      <li
        className="divider"
      />
      <li
        key="0"
      >
        <a
          className="location_menuitem"
          data-id="aid_locations_undefined"
          onClick={[Function]}
        />
      </li>
    </ul>
  </NavItem>
  <NavItem
    activeHref=""
    activeKey=""
    className=""
    id="vertical-spinner"
  >
    <Spinner
      className=""
      inline={false}
      inverse={true}
      loading={true}
      size="md"
    />
  </NavItem>
</Nav>
`;
